pipeline/processor/postprocessor.py

```diff
@@ -0,0 +1,634 @@
+
+    @staticmethod
+    def _classify_text_block_styling(text_block: TextBlock) -> Optional[str]:
+        """
+        Get text block styling, if present.
+
+        Args:
+            text_block:
+
+        Returns:
+
+        """
+        if not text_block["custom_attributes"]:
+            return None
+
+        if text_block["custom_attributes"].get("BaselineShift", 0) < 0:
+            return "subscript"
+        elif text_block["custom_attributes"].get("TextDecorationType") == "Underline":
+            return "underline"
+        elif text_block["custom_attributes"].get("TextPosition") == "Sup":
+            return "superscript"
+        else:
+            return None
+
+    @staticmethod
+    def _add_text_styling_markers(text: str, styling: str) -> str:
+        """
+        Add inline styling markers to text.
+
+        Args:
+            text: raw text without styling.
+            styling: styling to apply.
+
+        Returns:
+
+        """
+        leading_spaces = " " * (len(text) - len(text.lstrip(" ")))
+        trailing_spaces = " " * (len(text) - len(text.rstrip(" ")))
+
+        if styling == "subscript":
+            # Keep subscripts as they may be semantically relevant (as in CO2)
+            return f"{leading_spaces}<sub>{text.strip()}</sub>{trailing_spaces}"
+        elif styling == "superscript":
+            # Superscripts not semantically relevant, remove them.
+            return f"{leading_spaces + trailing_spaces}"
+        elif styling == "underline":
+            # TODO: What is the semantic relevance of underlines?
+            return f"{leading_spaces}<u>{text.strip()}</u>{trailing_spaces}"
+        else:
+            return text
+
+    def merge_text_blocks(self, text_blocks: List[dict]) -> dict:
+        """
+        Merge text blocks in the same semantic category (same path) that have been separated due to styling elements.
+
+        Args:
+            text_blocks:
+
+        Returns:
+            A new text block
+
+        """
+        all_coords = [tuple(text_block["coords"]) for text_block in text_blocks]
+        merged_coords = self._minimal_bounding_box(all_coords)
+
+        style_spans = []
+        merged_block_text = []
+        cumulative_block_len = 0
+        for text_block in text_blocks:
+            start_ix = cumulative_block_len
+            cumulative_block_len += sum([len(line) for line in text_block["text"]])
+            block_styling = self._classify_text_block_styling(text_block)
+
+            if merged_block_text == []:
+                merged_block_text = text_block["text"]
+            else:
+                merged_block_text[-1] = merged_block_text[-1] + text_block["text"][0]
+                merged_block_text += text_block["text"][1:]
+
+            # Append style metadata that will later be added to custom attributes for text block.
+            if block_styling:
+                last_style_ix = len(merged_block_text[0].strip()) - 1
+                style_spans.append(
```

Comment:
Spacy styling here, as discussed. Turned out to be much simpler. Do we want to do something similar for lists? At the moment, the I've formatted lists in 2 ways - with html -esque formatting in the text block itself and with pretty printing in the metadata.

Reply:
I think lists make sense as one line (string in `TextBlock.text` list) per bullet - how would you imagine using this styling approach for lists?

Lists were structured pretty well last time I saw the code (haven't checked this time though). As far as I can remember the only potentially outstanding issue was nested lists.

Reply:
@kdutia  Yes so I refactored it a bit. Nested lists are handled now with html-esque tags (so in my formatting there is
li1, li2, etc at the moment, and nesting is dealt with via tabulation and detabulation in the pprint function. Should be relatively easy to refactor the html-esque styling into something else with what I've got now if need be.

As far as the text goes, there is one line per bullet as you suggest, but I have something like <Lbl>(a)<\Lbl><Lbody>...<\Lbody>. ofc I could have normalised the labels but thought in some cases knowing an (a) from a bullet is useful, though though metadata on nesting level is probably enough.

There are still some issues the logic occasionally hits like deciding whether to interpret a span as a bullet or italics in a list body, as both cases occur. But since I kept finding issues like this that crop up occasionally, I've ignored them as I don't have true visibility on how common they are (quite hard to tell without manual work).

Reply:
Let's not worry about edge cases for now. We'll have a better grip on how prevalent they are when we run these postprocessing steps on the whole corpus, and we can do that in #391.

Maybe it's worth creating a github markdown or Notion page giving examples of inputs and outputs for each of the steps, with some of the details you've mentioned here? I get the gist of your comments but am finding it hard to visualise them without examples, which probably means we'll find these comments hard to decode in a couple of months too.

Reply:
Yep definitely agree, will set up a notion page.